################################################################################
SHORT DESCRIPTION: 
################################################################################
XSA-160 (http://xenbits.xen.org/xsa/advisory-160.html)
libxl leak of pv kernel and initrd on error

################################################################################
LONG DESCRIPTION: 
################################################################################
Source: http://xenbits.xen.org/xsa/advisory-160.html
Patches: xsa160-4.4.patch

when constructing a guest which is configured to use a pv bootloader
which runs as a userspace process in the toolstack domain
(e.g. pygrub) libxl creates a mapping of the files to be used as
kernel and initial ramdisk when building the guest domain.

however if building the domain subsequently fails these mappings would
not be released leading to a leak of virtual address space in the
calling process, as well as preventing the recovery of the temporary
disk files containing the kernel and initial ramdisk.

################################################################################
CHANGELOG 
################################################################################
Added in OpenXT, Xen 4.3.4 patch-queue: Eric Chanudet <chanudete@ainfosec.com>

################################################################################
PATCHES 
################################################################################
Index: xen-4.3.4/tools/libxl/libxl_create.c
===================================================================
--- xen-4.3.4.orig/tools/libxl/libxl_create.c	2015-03-19 16:08:36.000000000 +0100
+++ xen-4.3.4/tools/libxl/libxl_create.c	2015-12-18 18:09:49.188858815 +0100
@@ -1197,6 +1197,9 @@
     STATE_AO_GC(dcs->ao);
     libxl_domain_config *const d_config = dcs->guest_config;
 
+    libxl__file_reference_unmap(&dcs->build_state.pv_kernel);
+    libxl__file_reference_unmap(&dcs->build_state.pv_ramdisk);
+
     if (!rc && d_config->b_info.exec_ssidref)
         rc = xc_flask_relabel_domain(CTX->xch, dcs->guest_domid, d_config->b_info.exec_ssidref);
 
Index: xen-4.3.4/tools/libxl/libxl_dom.c
===================================================================
--- xen-4.3.4.orig/tools/libxl/libxl_dom.c	2015-12-14 13:24:27.655106526 +0100
+++ xen-4.3.4/tools/libxl/libxl_dom.c	2015-12-18 18:09:49.192192102 +0100
@@ -420,9 +420,6 @@
         state->store_mfn = xc_dom_p2m_host(dom, dom->xenstore_pfn);
     }
 
-    libxl__file_reference_unmap(&state->pv_kernel);
-    libxl__file_reference_unmap(&state->pv_ramdisk);
-
     ret = 0;
 out:
     xc_dom_release(dom);
